src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,117 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size = 0;
+    private T[] arr;
+
+    public ArrayList() {
+        arr = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size < arr.length) {
+            arr[size] = value;
+            size++;
+        } else {
+            resize();
+            arr[size] = value;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+        if (index == arr.length || size >= arr.length) {
+            resize();
+        }
 
+        T[] newArr = (T[]) new Object[arr.length];
+        System.arraycopy(arr, 0, newArr, 0, size);
+        newArr[index] = value;
+        System.arraycopy(arr, index, newArr, index + 1, size - index);
+        System.arraycopy(newArr, 0, arr, 0, newArr.length);
+        if (index >= 0 && index <= size) {
+            size++;
+        }
```

Comment:
The logic for shifting elements after inserting a new one should be extracted to a separate method to avoid code duplication and to follow the DRY principle.